### PR TITLE
HHH-10668 - Add SQL dialect for SQLite 3

### PR DESCRIPTION
**Pull request by gwenn**

### CODE REVIEW

**Comment by vladmihalcea** [comment]
Thanks for the PR. I'm going to review it.

**Comment by vladmihalcea** [comment]
Hi @gwenn ,

You should discuss the new Dialect with @sebersole. I asked about integrating this Dialect , and this is the [response that I got](https://www.mail-archive.com/hibernate-dev%40lists.jboss.org/msg14315.html). Basically, you need to add a JIRA issue and then the Dialect is going to be integrated only if there is demand from the community for it. 

Vlad

**Comment by luca-vercelli** [comment]
Are there any news about this issue? I am interested in it, too.
I think there actually **is** demand for this feature, as one can see [searching in github](https://github.com/search?utf8=%E2%9C%93&q=sqlite+dialect&type=).

**Comment by vladmihalcea** [comment]
@luca-vercelli I think you have a very good point. I asked on the mailing list if anyone opposes integrating it, and, otherwise, I'll integrate it this week.

**Comment by vladmihalcea** [comment]
@gwenn As discussed on the [mailing list](http://lists.jboss.org/pipermail/hibernate-dev/2017-March/016020.html), the only concern is related to Dialect maintainability on the long run.

Are you willing to contribute to the SQLite Dialect to accommodate upcoming features or fixing issues? If you agree, then we can integrate it in hibernate-core.

**Comment by gwenn** [comment]
Sorry,
I didn't reply for several reasons.

* For the third point: "The community demands it", I don't see how I can do anything.

* For the second point: "Someone or some entity is willing to maintain it *long term*",
I can help but the current implementation has many limitations (SQLite has a dynamic type system (column declared type does not mean anything except giving an hint on affinity), has a very limited ALTER TABLE command and JDBC drivers are not [compliant](https://docs.oracle.com/javase/8/docs/api/java/sql/Driver.html#jdbcCompliant--)).
So I am afraid that intractable issues will created.

And I don't think [Hwaci](http://www.hwaci.com) will be interested to maintain Java code.

* And I have limited internet access from my computer at work, and a single old laptop at home; I am not subscribed to any Hibernate topic. So testing/following the huge Hibernate ORM project is difficult to me.

To sum up, I am afraid you are asking too much of me (or I misunderstood) !

**Comment by vladmihalcea** [comment]
Thanks, @gwenn for your reply. I think we should postpone the integration until the SQLite JDBC Driver is complaint and we have someone to lead this Dialect integration.

Meanwhile, the SQLite Dialect can still be used as a 3rd party API.

**Comment by luca-vercelli** [comment]
Dear @gwenn and @vladmihalcea , I can understand your point, anyway I do not agree with you.

I know that JDBC drivers for SQLite are _bad_ so far, however this is their problem, not Hibernate's.
If you want, I could give "long term support" for this simple class (SQLiteDialect.java), not for JDBC driver. I think issues concerning the drivers must be sent to their developers.

The more, not all features of SQLite can be handled by Hibernate, but this is not a big problem IMHO. For example, Oracle's stored procedures are a great feature which is badly supported by Hibernate, anyway this is not a good reason for removing Oracle's dialect!

**Comment by vladmihalcea** [comment]
@luca-vercelli It's just a matter of resources. The Hibernate ORM is developed by a handful of people that need to work on new features, fix issues, answer questions on our forum and StackOverflow, write documentation, write blog posts, and so on. 

We can cover the most common RDBMS you'll find in a typical enterprise system, but our bandwidth is limited. So, when you have limited resources, you need to prioritize, which is exactly the reason why we cannot accept a new Dialect unless there is someone who is willing to take responsibility for it.

> If you want, I could give "long term support" for this simple class (SQLiteDialect.java)

That's good news. What we need besides this new dialect class `Dialect` can be summarized as follows:

- a new entry in `databases.gradle` for sqllite with a typical connection property configuration
- a test dependency in `libraries.gradle` for `sqlite-jdbc`
- in `build.gralde`, we also need the `testRuntime( libraries.sqllite)` to include `sqlite-jdbc` di=uring testing

To test how all tests run on the new SQLite dialect, just run:

    > gradle clean test -Pdb=sqllite

where `sqllite` is actually the key of the entry you added in `databases.gradle`.

@luca-vercelli Please send me a new Pull Request based on this one that adds the aformentioned changes, and which runs fine on SQLite. Then, I'll tested it and integrate it.
